### ISSUE TITLE
Credentials Setup Without Admin Access

### ISSUE THREAD

**Issue by eahefnawy** [serverless/serverless#3084]
Been researching hard on how we can finally drop the admin access "requirement" of the framework. This simply means that we need to figure out exactly what operations and permissions is needed by the framework and translate that into a valid IAM policy document. Admins will then use this IAM policy and manually attach it to their team to generate the api key/secret with just enough permissions, which won't include permissions to create IAM roles.

**But wait, doesn't the framework need to create IAM roles for the functions?** ... Correct, so that means if you choose the "limited credentials setup" instead of "admin setup", you're required to use the custom roles feature which we already support. That means team admins are also now responsible for creating that role manually.

So in summary, team admins will need to do two things manually:

* Attach the limited permissions policy to users in their team. They're free to edit it for further control.
* Create the role that will be assumed by the functions, either at the service level, or function level. The policy in this role minimally just include statements for creating logs.

This seems nice, and all we have to do is just figure out this minimum policy documents and document it.

## But how strict should this policy document be?
Ideally, it'd be so strict that it only allows permissions to perform on specific resources (ie. specific lambda functions that is used by the service). This is hard to accomplish because using the framework means always creating new resources (ie. creating new lambda functions), so you don't really know beforehand which resources you need access to. Here are some solutions to solve this issue:

### The Easy Solution: Do nothing
We can simply provide the minimum policy document **without defining specific resources**. So that means if the admin uses this document as is, users will have semi-admin access (they'll access all lambda functions in account for example, but not all AWS services). The admin is free to manually edit the document as needed based on service needs. This could be tedious though, since admins have to edit permissions for each new function or event source created...and of course custom resources.

### Better UX Solution: Plugin for everyone
So far we haven't really written any code to solve this admin access problem. But to make things a little easier for admins, we can introduce a new plugin that generates the minimum policy needed based on the config in `serverless.yml`.

Whenever the service structure has been changed, anyone can run `serverless permissions` or something, that will generate a json file with the minimum permissions based on that service. This specific command shouldn't need any permissions at all, but will still require api key/secret to get the AWS account ID with getUSer for the user to generate valid ARNs. Admins should run that command for every time the service change, and use the generated document manually with AWS console or CLI.

### Best UX Solution: Plugin for admins only
To take it one step further with UX, we can build a plugin for admins only, which means it'll assume that it's running on a machine with admin access to AWS. Instead of generating the policy to be used manually, it'll actually edit the IAM policy of the desired users to sync the permissions based on what's needed in `serverless.yml`.

The way it'd work at a high level is that an admin lists his team in `serverless.yml` and runs `serverless sync permissions`. The framework will then sync the minimum required permissions to the entire team. Again, this has to be done whenever the service change, or we can even provide a watcher feature.

# Conclusion
It may be an overkill to write a plugin to solve this issue, since the whole point of this solution is to give admins more control. Writing a plugin for admins means that the framework will still have admin access, although it'd be used only by admins. Also writing that kinda plugin might be a hard task, since we'll have to take into account all AWS resources. It'll be just like AWS Policy Generator, but specific to Serverless.

I personally think we should go with the easy solution first, and see if there's a need for further automation for admins. What you guys think?

cc | @serverless/team @serverless/vip

**Comment by eahefnawy** [comment]
Also, does anyone know of an open source tool that inspects a CloudFormation template and generates the needed policy document for that template? This goes beyond Serverless, and something like that has to exist! This could make writing the plugin much easier!

**Comment by pmuens** [comment]
Thank you for the in-depth research you've done here @eahefnawy 

I really like the idea of having a plugin which will generate the corresponding policies!
Right now I'm in favor of option 2 since option 1 and 3 seems to be more complex and needs more manual user interaction.

Or is there any possibility to reduce the tedious work when going with option 1?

**Comment by horike37** [comment]
I think also option 2 is good. It is moderately simple and easy to understand

**Comment by erikerikson** [comment]
TL;DR:
1. dimension: separate IAM CFT or not?
2. dimension: separate deployment IAM role?  if yes, specificity of rights?

Discussion:
Technically speaking, the "bring your own role" feature drops the "Admin Rights" requirement for the framework already.  As such, I believe what we're discussing here is how to facilitate workflows that allow the framework to assist users in managing those separated rights.  We are talking about use cases where users are willing and want to do more work in order to be more careful but as little extra work as possible.

I'd suggest there at least two important dimensions of increasing paranoia and/or maturity: "privilege escalation protection" and "separated deployment rights".

Dimension "Privilege Escalation Protection":
```
1. Whatevs.
2. Separate IAM from Everything Else (EE)
```

`1.` This option is great for agility and getting started or trying out the framework in a new throw-away account.  It is also great if your organization is mature in its cloud adoption and has the ability to evaluate and avoid privilege escalation.  This is the approach currently implemented.
`2.` This option is a great "poor person's" stop gap for avoiding privilege escalation.  The broad approach is to control who has the rights (or the scenario in which those rights are used) to alter IAM.

Concretely:
`input`: serverless.yml
`output.1`: Service CFT
`output.2`: IAM Roles CFT, IAM Resources CFT

Dimension "Separated Deployment Rights":
```
a) Whatevs on EE (no separate deployment role)
b) Generate a deployment IAM Role with:
   i) Broad Rights to Deploy EE Based on the Technologies Currently Used by the Service
   ii) Exact Rights to Deploy EE Based on the Resources Specified by the Service
```

`a)` This represents the case that users want to declare an externally managed IAM role for deployment.  Either they use their own rights or they have a set of credentials used by their CI/CD solution and it is the rights of that role under which deployments occur.  Arbitrary control is possible.

`b.i)` This represents the use case that users want to broadly control the technologies that can be used by a service.  This way developers can, for example, create any number of Lambdas and API Gateway Endpoints within a service but would need to use the collaborative/admin process to use any new technology in their service.  For example, adding the rights to create a Kinesis Stream would require that process.  I'm tempted to omit this use case.

`b.ii)` This represents the use case that users want to tightly lock down what can be deployed in a service.  This could give you a certain measure of peace of mind for a case where the user's deployment role is compromised or the user's CI/CD system is owned.  Perhaps, instead, you have regulatory requirements that demand auditing for certain change management processes.  Under this scenario users would have to follow the collaborative/admin process whenever creating new resources but could iterate on the configuration or implementation (i.e. Lambda code) of existing resources.

Concretely:
`input`: serverless.yml
`output.a`: Service CFT
`output.b`: Deployment Role CFT, Service CFT (`.i` narrow rights, `.ii` exact rights).  Deployment Role used by framework via [TemporaryCredentials](http://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/TemporaryCredentials.html).

All Together, Concretely:
`input`: serverless.yml
`output.1+a`: Service CFT
`output.2+a`: IAM CFT, Resources CFT
`output.1+b`: Deployment Role CFT, Service CFT (`.i` narrow rights, `.ii` exact rights)
`output.2+b`: IAM Deployment CFT, IAM CFT, Resource Deployment CFT, Resource CFT (`.i` narrow rights, `.ii` exact rights)

Commentary:
From a prioritization standpoint, I'd suggest this is far less important than increasing the specificity of IAM rights provided to our Lambdas.  See [Gone In 60 Milliseconds](https://media.ccc.de/v/33c3-7865-gone_in_60_milliseconds#video) for some reasons why we might care.  If SLS was great at being extremely specific, it'd go a long way towards the reduction of security threats, blast zones, and accidents.

**Comment by KevinColemanInc** [comment]
Is this dead?  My boss won't let me have a full access AWS IAM user and I am struggling to figure out what the min permissions are.

**Comment by horike37** [comment]
@KevinColemanInc 
The discussion doesn't seem to be getting anywhere.
However, you want to narrow more permissions. You can see the #1439 issue which is to do that.
Hope that would be some of help for you.

**Comment by kenjinp** [comment]
It seems like the collective opinion is to shrug their shoulders at this enormous issue 🤷‍♀️

**Comment by erikerikson** [comment]
Unfortunately, we are all busy, there are a lot of interlocking concerns in the solution, and the solution implies unbounded contribution as IAM changes (ignoring that IAM is obscure and imperfectly documented).  Many users here in my enterprise care deeply about this.  On my team, we currently do it manually (i.e. generating absolutely least privilege deployment and runtime roles).  It is far more painful than it need be.

Serverless Team (@eahefnawy, @pmuens, et. al.): you might get value out of reaching out to the PurSec team.  They've done some awesome work in this space.

**Comment by pmuens** [comment]
Thanks for commenting @kenjinp and thanks for the answer @erikerikson 👍

I can only echo what Erik said about proper IAM setup. For everyone else looking here we'd definitely recommend to check out this plugin https://github.com/puresec/serverless-puresec-cli by the PureSec guys! It's super nice and will help you creating proper IAM roles.

**Comment by erikerikson** [comment]
💯% awesome!  I hadn't realized that was public yet. 

Awesome tool but do note that it currently generates runtime roles, not the deployment roles we're discussing in this context.

**Comment by Ray-B** [comment]
I'm probably not the only one feeling ambivalence about this. I'm gracious for the value of serverless, and the hard work done, and simultaneously irritated by the lack of traction this _fundamental_ issue _appears_ to be receiving in public. No doubt it is frustrating for everyone.

After many hours of lateral jumping between issues and comments, I'm still not sure what the serverless' team's official recommendation here is - can anyone offer clarification or point to the documentation?

Is there some documented example set of AWS Permissions required by sls for the basic use case of creating a lambda, or is the process just trial an error? I've tried various permutations from the comments, but it seemed to be a game of whack-a-mole.

I see https://github.com/puresec/serverless-puresec-cli mentioned, but this looks to be geared towards the infrastructure sls is responsible for.

I also saw @dancrumb generously created: https://github.com/dancrumb/generator-serverless-policy Is this still the best viable option?

https://github.com/apex/up/issues/277 Is probably the best start to making some progress, but after over a year and a half, it seems there is limited interest in pushing additional documentation forward.

If there is a desire/plan to work on improving this, how can the community assist/what is needed?
